File: mobile/ios-driver-app/Tests/E2E/SingleOrderE2ETest.swift
/**
 * E2E Test: Single Order Flow
 * 
 * Tests the complete flow of accepting and completing a single order
 * Verifies earnings calculation matches driverEarningsService
 */

import XCTest
@testable import SpeedyVanDriver

final class SingleOrderE2ETest: XCTestCase {
    
    var app: XCUIApplication!
    
    override func setUpWithError() throws {
        continueAfterFailure = false
        app = XCUIApplication()
        app.launchArguments = ["UI-Testing"]
        app.launch()
    }
    
    override func tearDownWithError() throws {
        app = nil
    }
    
    // MARK: - Test Single Order Flow
    
    func testSingleOrderCompleteFlow() throws {
        // 1. Login
        loginAsTestDriver()
        
        // 2. Navigate to Available Jobs
        let availableJobsTab = app.tabBars.buttons["Available"]
        XCTAssertTrue(availableJobsTab.waitForExistence(timeout: 5))
        availableJobsTab.tap()
        
        // 3. Wait for jobs to load
        let firstJobCard = app.scrollViews.otherElements.containing(.staticText, identifier: "job-card").element
        XCTAssertTrue(firstJobCard.waitForExistence(timeout: 10))
        
        // 4. Verify earnings display (no percentage!)
        let earningsLabel = firstJobCard.staticTexts.matching(identifier: "estimated-earnings").element
        XCTAssertTrue(earningsLabel.exists)
        
        let earningsText = earningsLabel.label
        XCTAssertTrue(earningsText.contains("£"))
        XCTAssertFalse(earningsText.contains("%"), "Earnings should NOT show percentage")
        
        // Extract earnings amount
        let earningsAmount = extractAmount(from: earningsText)
        XCTAssertGreaterThan(earningsAmount, 0, "Earnings must be positive")
        
        // 5. Tap to view job details
        firstJobCard.tap()
        
        // 6. Verify Job Detail Screen
        let jobDetailTitle = app.navigationBars["Job Details"]
        XCTAssertTrue(jobDetailTitle.waitForExistence(timeout: 5))
        
        // 7. Verify earnings breakdown
        let baseFareLabel = app.staticTexts["Base Fare"]
        let mileageFeeLabel = app.staticTexts["Mileage Fee"]
        let timeFeeLabel = app.staticTexts["Time Fee"]
        
        XCTAssertTrue(baseFareLabel.exists, "Base fare should be shown")
        XCTAssertTrue(mileageFeeLabel.exists, "Mileage fee should be shown")
        XCTAssertTrue(timeFeeLabel.exists, "Time fee should be shown")
        
        // 8. Accept the job
        let acceptButton = app.buttons["Accept Job"]
        XCTAssertTrue(acceptButton.exists)
        acceptButton.tap()
        
        // 9. Confirm acceptance
        let confirmButton = app.alerts.buttons["Confirm"]
        if confirmButton.waitForExistence(timeout: 2) {
            confirmButton.tap()
        }
        
        // 10. Verify navigation to Active Jobs
        let activeJobsTab = app.tabBars.buttons["Active"]
        XCTAssertTrue(activeJobsTab.waitForExistence(timeout: 5))
        activeJobsTab.tap()
        
        // 11. Verify job appears in active list
        let activeJobCard = app.scrollViews.otherElements.containing(.staticText, identifier: "active-job-card").element
        XCTAssertTrue(activeJobCard.waitForExistence(timeout: 5))
        
        // 12. Start the job
        activeJobCard.tap()
        let startButton = app.buttons["Start Job"]
        XCTAssertTrue(startButton.exists)
        startButton.tap()
        
        // 13. Verify live tracking
        let mapView = app.otherElements["map-view"]
        XCTAssertTrue(mapView.waitForExistence(timeout: 5))
        
        let currentLocationMarker = app.otherElements["current-location-marker"]
        XCTAssertTrue(currentLocationMarker.exists, "Current location should be shown on map")
        
        // 14. Arrive at pickup
        let arrivePickupButton = app.buttons["Arrive at Pickup"]
        XCTAssertTrue(arrivePickupButton.waitForExistence(timeout: 5))
        arrivePickupButton.tap()
        
        // 15. Start delivery
        let startDeliveryButton = app.buttons["Start Delivery"]
        XCTAssertTrue(startDeliveryButton.waitForExistence(timeout: 5))
        startDeliveryButton.tap()
        
        // 16. Arrive at dropoff
        let arriveDropoffButton = app.buttons["Arrive at Dropoff"]
        XCTAssertTrue(arriveDropoffButton.waitForExistence(timeout: 5))
        arriveDropoffButton.tap()
        
        // 17. Complete delivery
        let completeButton = app.buttons["Complete Delivery"]
        XCTAssertTrue(completeButton.waitForExistence(timeout: 5))
        completeButton.tap()
        
        // 18. Verify completion screen
        let completionTitle = app.navigationBars["Job Completion"]
        XCTAssertTrue(completionTitle.waitForExistence(timeout: 5))
        
        // 19. Take photo proof
        let photoButton = app.buttons["Take Photo"]
        XCTAssertTrue(photoButton.exists)
        photoButton.tap()
        
        // Simulate camera (in UI test, use mock)
        let usePhotoButton = app.buttons["Use Photo"]
        if usePhotoButton.waitForExistence(timeout: 3) {
            usePhotoButton.tap()
        }
        
        // 20. Get signature
        let signatureButton = app.buttons["Get Signature"]
        XCTAssertTrue(signatureButton.exists)
        signatureButton.tap()
        
        // Draw signature (simulate)
        let signatureCanvas = app.otherElements["signature-canvas"]
        XCTAssertTrue(signatureCanvas.exists)
        signatureCanvas.swipeRight()
        
        let doneSignatureButton = app.buttons["Done"]
        XCTAssertTrue(doneSignatureButton.exists)
        doneSignatureButton.tap()
        
        // 21. Submit completion
        let submitButton = app.buttons["Submit"]
        XCTAssertTrue(submitButton.exists)
        submitButton.tap()
        
        // 22. Verify earnings summary
        let earningSummaryTitle = app.staticTexts["Earnings Summary"]
        XCTAssertTrue(earningSummaryTitle.waitForExistence(timeout: 10))
        
        // 23. Verify final earnings
        let finalEarningsLabel = app.staticTexts.matching(identifier: "final-earnings").element
        XCTAssertTrue(finalEarningsLabel.exists)
        
        let finalEarningsText = finalEarningsLabel.label
        let finalAmount = extractAmount(from: finalEarningsText)
        
        // 24. Verify earnings are reasonable
        XCTAssertGreaterThan(finalAmount, 20.0, "Minimum earnings should be > £20")
        XCTAssertLessThan(finalAmount, 500.0, "Daily cap is £500")
        
        // 25. Verify no platform fee shown
        let platformFeeLabel = app.staticTexts["Platform Fee"]
        XCTAssertFalse(platformFeeLabel.exists, "Platform fee should NOT be shown")
        
        // 26. Verify breakdown components
        let baseFareAmount = extractAmount(from: app.staticTexts["base-fare-amount"].label)
        let mileageAmount = extractAmount(from: app.staticTexts["mileage-amount"].label)
        let timeAmount = extractAmount(from: app.staticTexts["time-amount"].label)
        
        XCTAssertGreaterThan(baseFareAmount, 0, "Base fare must be positive")
        XCTAssertGreaterThanOrEqual(mileageAmount, 0, "Mileage fee must be non-negative")
        XCTAssertGreaterThanOrEqual(timeAmount, 0, "Time fee must be non-negative")
        
        // 27. Verify total matches sum
        let calculatedTotal = baseFareAmount + mileageAmount + timeAmount
        XCTAssertEqual(finalAmount, calculatedTotal, accuracy: 0.01, "Total should match sum of components")
        
        // 28. Done
        let doneButton = app.buttons["Done"]
        XCTAssertTrue(doneButton.exists)
        doneButton.tap()
        
        // 29. Verify back to dashboard
        let dashboardTitle = app.navigationBars["Dashboard"]
        XCTAssertTrue(dashboardTitle.waitForExistence(timeout: 5))
        
        print("✅ Single Order E2E Test Passed!")
    }
    
    // MARK: - Test Earnings Calculation Accuracy
    
    func testEarningsCalculationAccuracy() throws {
        // This test verifies that the iOS app calculates earnings correctly
        // by comparing with the expected formula:
        // Earnings = Base Fare + (Distance × £0.55) + (Duration × £0.15) + Bonuses - Penalties
        
        loginAsTestDriver()
        
        // Navigate to a test job with known parameters
        let availableJobsTab = app.tabBars.buttons["Available"]
        availableJobsTab.tap()
        
        // Find test job (in UI test, we'd use a mock job with known values)
        let testJobCard = app.scrollViews.otherElements.containing(.staticText, identifier: "test-job-card").element
        XCTAssertTrue(testJobCard.waitForExistence(timeout: 10))
        
        // Extract job details
        let distanceLabel = testJobCard.staticTexts.matching(identifier: "distance").element
        let distanceText = distanceLabel.label
        let distance = extractNumber(from: distanceText) // e.g., "12.5 miles" → 12.5
        
        let durationLabel = testJobCard.staticTexts.matching(identifier: "duration").element
        let durationText = durationLabel.label
        let duration = extractNumber(from: durationText) // e.g., "45 min" → 45
        
        let earningsLabel = testJobCard.staticTexts.matching(identifier: "estimated-earnings").element
        let earningsText = earningsLabel.label
        let displayedEarnings = extractAmount(from: earningsText)
        
        // Calculate expected earnings
        let baseFare = 25.0
        let perMileFee = 0.55
        let perMinuteFee = 0.15
        
        let expectedEarnings = baseFare + (distance * perMileFee) + (duration * perMinuteFee)
        
        // Verify accuracy (allow 1% tolerance for rounding)
        let tolerance = expectedEarnings * 0.01
        XCTAssertEqual(displayedEarnings, expectedEarnings, accuracy: tolerance, 
                      "Displayed earnings should match calculated earnings")
        
        print("✅ Earnings Calculation Accuracy Test Passed!")
        print("   Distance: \(distance) miles")
        print("   Duration: \(duration) minutes")
        print("   Expected: £\(expectedEarnings)")
        print("   Displayed: £\(displayedEarnings)")
    }
    
    // MARK: - Test Daily Cap Enforcement
    
    func testDailyCap Enforcement() throws {
        loginAsTestDriver()
        
        // Navigate to Earnings
        let earningsTab = app.tabBars.buttons["Earnings"]
        earningsTab.tap()
        
        // Check today's total
        let todayTotalLabel = app.staticTexts["today-total"]
        XCTAssertTrue(todayTotalLabel.waitForExistence(timeout: 5))
        
        let todayTotal = extractAmount(from: todayTotalLabel.label)
        
        // Verify cap warning if approaching £500
        if todayTotal >= 450.0 {
            let capWarning = app.staticTexts["cap-warning"]
            XCTAssertTrue(capWarning.exists, "Warning should show when approaching daily cap")
        }
        
        // Verify cannot exceed £500
        XCTAssertLessThanOrEqual(todayTotal, 500.0, "Daily total cannot exceed £500")
        
        print("✅ Daily Cap Enforcement Test Passed!")
        print("   Today's Total: £\(todayTotal)")
    }
    
    // MARK: - Helper Methods
    
    private func loginAsTestDriver() {
        let emailField = app.textFields["email"]
        XCTAssertTrue(emailField.waitForExistence(timeout: 5))
        emailField.tap()
        emailField.typeText("test.driver@speedyvan.co.uk")
        
        let passwordField = app.secureTextFields["password"]
        passwordField.tap()
        passwordField.typeText("TestDriver123!")
        
        let loginButton = app.buttons["Login"]
        loginButton.tap()
        
        // Wait for dashboard
        let dashboardTitle = app.navigationBars["Dashboard"]
        XCTAssertTrue(dashboardTitle.waitForExistence(timeout: 10))
    }
    
    private func extractAmount(from text: String) -> Double {
        // Extract amount from text like "£45.50" or "Earnings: £45.50"
        let pattern = "£([0-9]+\\.?[0-9]*)"
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              let range = Range(match.range(at: 1), in: text) else {
            return 0.0
        }
        return Double(text[range]) ?? 0.0
    }
    
    private func extractNumber(from text: String) -> Double {
        // Extract number from text like "12.5 miles" or "45 min"
        let pattern = "([0-9]+\\.?[0-9]*)"
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              let range = Range(match.range(at: 1), in: text) else {
            return 0.0
        }
        return Double(text[range]) ?? 0.0
    }
}

File: mobile/expo-driver-app/e2e/singleOrder.e2e.test.ts
/**
 * E2E Test: Single Order Flow (Android/Expo)
 * 
 * Tests the complete flow of accepting and completing a single order
 * Verifies earnings calculation matches driverEarningsService
 * 
 * Run with: detox test --configuration android.emu.debug
 */

import { device, element, by, expect as detoxExpect, waitFor } from 'detox';

describe('Single Order E2E Test', () => {
  beforeAll(async () => {
    await device.launchApp({
      newInstance: true,
      permissions: { location: 'always', camera: 'YES', photos: 'YES' },
    });
  });

  beforeEach(async () => {
    await device.reloadReactNative();
  });

  // MARK: - Test Single Order Complete Flow

  it('should complete single order flow successfully', async () => {
    // 1. Login
    await loginAsTestDriver();

    // 2. Navigate to Available Jobs
    await element(by.id('tab-available')).tap();

    // 3. Wait for jobs to load
    await waitFor(element(by.id('job-card-0')))
      .toBeVisible()
      .withTimeout(10000);

    // 4. Verify earnings display (no percentage!)
    const earningsElement = element(by.id('job-earnings-0'));
    await detoxExpect(earningsElement).toBeVisible();

    // Get earnings text
    const earningsAttributes = await earningsElement.getAttributes();
    const earningsText = earningsAttributes.text || '';
    
    detoxExpect(earningsText).toContain('£');
    detoxExpect(earningsText).not.toContain('%');

    // Extract and verify amount
    const earningsAmount = extractAmount(earningsText);
    detoxExpect(earningsAmount).toBeGreaterThan(0);

    // 5. Tap to view job details
    await element(by.id('job-card-0')).tap();

    // 6. Verify Job Detail Screen
    await waitFor(element(by.text('Job Details')))
      .toBeVisible()
      .withTimeout(5000);

    // 7. Verify earnings breakdown
    await detoxExpect(element(by.text('Base Fare'))).toBeVisible();
    await detoxExpect(element(by.text('Mileage Fee'))).toBeVisible();
    await detoxExpect(element(by.text('Time Fee'))).toBeVisible();

    // 8. Accept the job
    await element(by.id('accept-job-button')).tap();

    // 9. Confirm acceptance
    await waitFor(element(by.text('Confirm')))
      .toBeVisible()
      .withTimeout(2000);
    await element(by.text('Confirm')).tap();

    // 10. Verify navigation to Active Jobs
    await waitFor(element(by.id('tab-active')))
      .toBeVisible()
      .withTimeout(5000);
    await element(by.id('tab-active')).tap();

    // 11. Verify job appears in active list
    await waitFor(element(by.id('active-job-card-0')))
      .toBeVisible()
      .withTimeout(5000);

    // 12. Start the job
    await element(by.id('active-job-card-0')).tap();
    await element(by.id('start-job-button')).tap();

    // 13. Verify live tracking
    await waitFor(element(by.id('map-view')))
      .toBeVisible()
      .withTimeout(5000);
    await detoxExpect(element(by.id('current-location-marker'))).toBeVisible();

    // 14. Arrive at pickup
    await waitFor(element(by.id('arrive-pickup-button')))
      .toBeVisible()
      .withTimeout(5000);
    await element(by.id('arrive-pickup-button')).tap();

    // 15. Start delivery
    await waitFor(element(by.id('start-delivery-button')))
      .toBeVisible()
      .withTimeout(5000);
    await element(by.id('start-delivery-button')).tap();

    // 16. Arrive at dropoff
    await waitFor(element(by.id('arrive-dropoff-button')))
      .toBeVisible()
      .withTimeout(5000);
    await element(by.id('arrive-dropoff-button')).tap();

    // 17. Complete delivery
    await waitFor(element(by.id('complete-delivery-button')))
      .toBeVisible()
      .withTimeout(5000);
    await element(by.id('complete-delivery-button')).tap();

    // 18. Verify completion screen
    await waitFor(element(by.text('Job Completion')))
      .toBeVisible()
      .withTimeout(5000);

    // 19. Take photo proof
    await element(by.id('take-photo-button')).tap();
    
    // Simulate camera (in test, use mock)
    await waitFor(element(by.id('use-photo-button')))
      .toBeVisible()
      .withTimeout(3000);
    await element(by.id('use-photo-button')).tap();

    // 20. Get signature
    await element(by.id('get-signature-button')).tap();
    
    // Draw signature (simulate)
    await element(by.id('signature-canvas')).swipe('right', 'fast', 0.5);
    await element(by.id('done-signature-button')).tap();

    // 21. Submit completion
    await element(by.id('submit-completion-button')).tap();

    // 22. Verify earnings summary
    await waitFor(element(by.text('Earnings Summary')))
      .toBeVisible()
      .withTimeout(10000);

    // 23. Verify final earnings
    const finalEarningsElement = element(by.id('final-earnings'));
    await detoxExpect(finalEarningsElement).toBeVisible();

    const finalEarningsAttributes = await finalEarningsElement.getAttributes();
    const finalEarningsText = finalEarningsAttributes.text || '';
    const finalAmount = extractAmount(finalEarningsText);

    // 24. Verify earnings are reasonable
    detoxExpect(finalAmount).toBeGreaterThan(20.0);
    detoxExpect(finalAmount).toBeLessThan(500.0);

    // 25. Verify no platform fee shown
    await detoxExpect(element(by.text('Platform Fee'))).not.toBeVisible();

    // 26. Verify breakdown components
    const baseFareElement = element(by.id('base-fare-amount'));
    const mileageElement = element(by.id('mileage-amount'));
    const timeElement = element(by.id('time-amount'));

    await detoxExpect(baseFareElement).toBeVisible();
    await detoxExpect(mileageElement).toBeVisible();
    await detoxExpect(timeElement).toBeVisible();

    const baseFareAttrs = await baseFareElement.getAttributes();
    const mileageAttrs = await mileageElement.getAttributes();
    const timeAttrs = await timeElement.getAttributes();

    const baseFareAmount = extractAmount(baseFareAttrs.text || '');
    const mileageAmount = extractAmount(mileageAttrs.text || '');
    const timeAmount = extractAmount(timeAttrs.text || '');

    detoxExpect(baseFareAmount).toBeGreaterThan(0);
    detoxExpect(mileageAmount).toBeGreaterThanOrEqual(0);
    detoxExpect(timeAmount).toBeGreaterThanOrEqual(0);

    // 27. Verify total matches sum
    const calculatedTotal = baseFareAmount + mileageAmount + timeAmount;
    const tolerance = 0.01;
    detoxExpect(Math.abs(finalAmount - calculatedTotal)).toBeLessThan(tolerance);

    // 28. Done
    await element(by.id('done-button')).tap();

    // 29. Verify back to dashboard
    await waitFor(element(by.text('Dashboard')))
      .toBeVisible()
      .withTimeout(5000);

    console.log('✅ Single Order E2E Test Passed!');
  });

  // MARK: - Test Earnings Calculation Accuracy

  it('should calculate earnings accurately', async () => {
    await loginAsTestDriver();

    // Navigate to available jobs
    await element(by.id('tab-available')).tap();

    // Find test job with known parameters
    await waitFor(element(by.id('test-job-card')))
      .toBeVisible()
      .withTimeout(10000);

    // Extract job details
    const distanceElement = element(by.id('test-job-distance'));
    const durationElement = element(by.id('test-job-duration'));
    const earningsElement = element(by.id('test-job-earnings'));

    const distanceAttrs = await distanceElement.getAttributes();
    const durationAttrs = await durationElement.getAttributes();
    const earningsAttrs = await earningsElement.getAttributes();

    const distance = extractNumber(distanceAttrs.text || ''); // e.g., "12.5 miles"
    const duration = extractNumber(durationAttrs.text || ''); // e.g., "45 min"
    const displayedEarnings = extractAmount(earningsAttrs.text || '');

    // Calculate expected earnings
    const baseFare = 25.0;
    const perMileFee = 0.55;
    const perMinuteFee = 0.15;

    const expectedEarnings = baseFare + (distance * perMileFee) + (duration * perMinuteFee);

    // Verify accuracy (allow 1% tolerance)
    const tolerance = expectedEarnings * 0.01;
    detoxExpect(Math.abs(displayedEarnings - expectedEarnings)).toBeLessThan(tolerance);

    console.log('✅ Earnings Calculation Accuracy Test Passed!');
    console.log(`   Distance: ${distance} miles`);
    console.log(`   Duration: ${duration} minutes`);
    console.log(`   Expected: £${expectedEarnings.toFixed(2)}`);
    console.log(`   Displayed: £${displayedEarnings.toFixed(2)}`);
  });

  // MARK: - Test Daily Cap Enforcement

  it('should enforce daily cap of £500', async () => {
    await loginAsTestDriver();

    // Navigate to Earnings
    await element(by.id('tab-earnings')).tap();

    // Check today's total
    await waitFor(element(by.id('today-total')))
      .toBeVisible()
      .withTimeout(5000);

    const todayTotalElement = element(by.id('today-total'));
    const todayTotalAttrs = await todayTotalElement.getAttributes();
    const todayTotal = extractAmount(todayTotalAttrs.text || '');

    // Verify cap warning if approaching £500
    if (todayTotal >= 450.0) {
      await detoxExpect(element(by.id('cap-warning'))).toBeVisible();
    }

    // Verify cannot exceed £500
    detoxExpect(todayTotal).toBeLessThanOrEqual(500.0);

    console.log('✅ Daily Cap Enforcement Test Passed!');
    console.log(`   Today's Total: £${todayTotal.toFixed(2)}`);
  });

  // MARK: - Helper Functions

  async function loginAsTestDriver() {
    await element(by.id('email-input')).typeText('test.driver@speedyvan.co.uk');
    await element(by.id('password-input')).typeText('TestDriver123!');
    await element(by.id('login-button')).tap();

    await waitFor(element(by.text('Dashboard')))
      .toBeVisible()
      .withTimeout(10000);
  }

  function extractAmount(text: string): number {
    const match = text.match(/£([0-9]+\.?[0-9]*)/);
    return match ? parseFloat(match[1]) : 0.0;
  }

  function extractNumber(text: string): number {
    const match = text.match(/([0-9]+\.?[0-9]*)/);
    return match ? parseFloat(match[1]) : 0.0;
  }
});

